### PR TITLE
[Bugfix][Spec Decode] Properly share dpsk-v4 topk_indices_buffer.

### PR DESCRIPTION
**Pull request by KungYork**
## Purpose

When using DeepSeek-V4 (or any sparse-attention model that exposes `topk_indices_buffer`) together with an MTP draft model for speculative decoding, the existing code in `vllm/v1/spec_decode/llm_base_proposer.py` looks like it shares the
target model's `topk_indices_buffer` with the draft, but in practice the share is a no-op:
```python
if hasattr(target_language_model.model, "topk_indices_buffer"):
    if hasattr(self.model.model, "topk_indices_buffer"):
        del self.model.model.topk_indices_buffer
    self.model.model.topk_indices_buffer = (
        target_language_model.model.topk_indices_buffer
    )
```
The buffer is allocated once in the draft model's __init__ (`DeepSeekV4MultiTokenPredictor.__init__`) and then passed by reference down to every submodule that consumes it:

`DeepSeekV4MultiTokenPredictor.topk_indices_buffer` (top-level)
`DeepseekV4MultiHeadLatentAttentionWrapper.topk_indices_buffer`
`DeepseekV4MLAAttention.topk_indices_buffer`
`DeepseekV4Indexer.topk_indices_buffer`
`SparseAttnIndexer.topk_indices_buffer` (the one actually read by torch.ops.vllm.sparse_attn_indexer at forward time)

Each of these stores its own attribute binding to the same Python tensor object. Re-assigning the top-level attribute therefore does not update the references already captured at construction time: the indexer ops keep writing into the draft's original buffer, and the original allocation is not freed because the indexer/MLA submodules still hold strong references to it.

Net effect: the draft model carries an extra max_num_batched_tokens * index_topk * sizeof(int32) GPU buffer that serves no purpose beyond what the target's buffer already provides.

## AI Assistance Disclosure
This change was developed with AI assistance (Cursor + Claude). I reviewed every changed line, ran the tests listed above, and take responsibility for the change end-to-end as required by AGENTS.md.

### CODE REVIEW

**Comment by benchislett** [comment]
Practically, this saves <100 MB. Low priority fix.

**Comment by benchislett** [comment]
Please reply to bot comment.

File: vllm/v1/spec_decode/llm_base_proposer.py [hunk 1]
```diff
@@ -1543,14 +1543,13 @@ def _maybe_share_lm_head(self, target_language_model: nn.Module) -> None:
...
-            self.model.model.topk_indices_buffer = (
-                target_language_model.model.topk_indices_buffer
-            )
-            logger.info(
-                "Detected MTP model with topk_indices_buffer. "
-                "Sharing target model topk_indices_buffer with the draft model."
-            )
+                self.model.model.topk_indices_buffer.set_(
+                    target_language_model.model.topk_indices_buffer
+                )
```
[the remark this answers is not in the document]
Comment (benchislett) [reply vllm/v1/spec_decode/llm_base_proposer.py]: Why is `set_` needed?

[the remark this answers is not in the document]
Comment (KungYork) [reply vllm/v1/spec_decode/llm_base_proposer.py]: `set_` is used make draft model indexer use the same topk memory buffer as main model indexer. I am also wondering whether there is a better way to achieve this.

**Comment by KungYork** [comment]
Yes, this patch does not save much memory. However, the output buffer used by `fp8_fp4_mqa_logits` is quite larger(typically `chunk_size` * `kv_len` * `output_bits`), and it is not currently designed for reuse. In high-memory-occupancy scenarios, this can repeatedly trigger allocations from the CUDA caching allocator. I think the same issue could arise again in the near future if this output memory is reused as a buffer in a similar way.

File: vllm/v1/spec_decode/llm_base_proposer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (benchislett) [reply vllm/v1/spec_decode/llm_base_proposer.py]: oh, my bad. I thought it was building a python `set` or something

**Review by benchislett — APPROVED** [review]
LGTM. Please run a basic accuracy test (GSM8k) on DSV4-Flash with MTP if you are able. If not, I can run it for you.
